Converted snippets to listings in Ch 9

elizabeth-pauley opened the pull request.

elizabeth-pauley committed:
Converted snippets to listings in Ch 9

BenjaminMichaelis commented:
Do you know how to point the word documents in the tooling repo to this commit so we can make sure the word doc parses?

@Keboo i would think it probably doesn't hurt to run this commit through the tooling? Thoughts?

Keboo commented:
@BenjaminMichaelis we already know it is going to fail when the word document runs through it. We tested locally and there will need to be document changes to go along with this. So this PR is going to sit until those get done.

elizabeth-pauley committed:
Added 2 new listings, new tests for recently added listings

elizabeth-pauley committed:
Updated test method names

Keboo commented on src/Chapter09.Tests/Listing09.02.Tests.cs:
```diff
@@ -1,15 +1,29 @@
+    public void Equals_CopyWithSameDegreesMinutesSeconds_AreEqual()
+    {
+        Angle angle1 = new(0, 0, 0);
+        Angle angle2 = new(0, 0, 0);
```
This works, but a slightly better option is to avoid using default values (such as 0 for numeric types) when you want to test their values since it is impossible to tell if the constructor is using the value you passed in or if the values are just defaulting to 0.

```suggestion
        Angle angle1 = new(1, 2, 3);
        Angle angle2 = new(1, 2, 3);
```

Keboo commented on src/Chapter09.Tests/Listing09.02.Tests.cs:
```diff
@@ -1,15 +1,29 @@
+    [TestMethod]
+    public void With_CopyWithSameDegreesMinutesSeconds_AreEqual()
+    {
+        Angle angle1 = new(0, 0, 0);
```
```suggestion
        Angle angle1 = new(1, 2, 3);
```

Keboo commented on src/Chapter09.Tests/Listing09.06.EquivalentRecordClassCode.Tests.cs:
```diff
@@ -1,23 +1,38 @@
 {
     [TestMethod]
-    public void EqualityContract_GetType()
+    public void With_Coordinate_IsReadOnly()
```
```suggestion
    public void With_Coordinate_AreEqual()
```

Keboo commented on src/Chapter09.Tests/Listing09.14.RecordDeconstructors.Tests.cs:
```diff
@@ -0,0 +1,24 @@
+        // The constructor is generated using positional parameters
+        Angle angle = new(degrees, minutes, seconds);
+
+        angle.Deconstruct(out int d, out int m, out int s);
```
nit: We generally only use abbreviations when they are obvious.
```suggestion
        angle.Deconstruct(out int degrees, out int minutes, out int seconds);
```

Keboo commented on src/Chapter09.Tests/Listing09.15.RecordDeconstructorsAndPatternMatching.Tests.cs:
```diff
@@ -0,0 +1,26 @@
+    [TestMethod]
+    public void Deconstructor_CorrespondingParameters_EnablePatternMatching()
+    {
+        (int degrees, int minutes, int seconds, string name) = (
+            90, 0, 0, null);
+
+        // The constructor is generated using positional parameters
+        Angle angle = new(degrees, minutes, seconds, name);
```
Though a clever way of doing multiple variable assignment. I think it might be clearer to simply pass these directly to the constructor.
```suggestion
        // The constructor is generated using positional parameters
        Angle angle = new(90, 0, 0, null);
```

Keboo commented on src/Chapter09.Tests/Listing09.15.RecordDeconstructorsAndPatternMatching.Tests.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            // ..
+            Assert.AreEqual(angle, angleData);
+        }
```
We should handle the else case, or rework this to not need the if-condition like we talked about yesterday.

Keboo commented on src/Chapter09.Tests/Listing09.22.OverridingGetHashCodeUsingCombine.Tests.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        Angle angle1 = new Angle(120, 20, 20);
+
+        int hashcode1 = angle1.GetHashCode();
+        int hashcode2 = HashCode.Combine(120, 20, 20);
+
+        Assert.IsTrue(hashcode1 == hashcode2);
```
nit: Casing
```suggestion
        int hashCode1 = angle1.GetHashCode();
        int hashCode2 = HashCode.Combine(120, 20, 20);

        Assert.IsTrue(hashCode1 == hashCode2);
```

Keboo commented on src/Chapter09.Tests/Listing09.22.OverridingGetHashCodeUsingCombine.Tests.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        Angle angle1 = new Angle(120, 20, 20);
+
+        int hashcode1 = angle1.GetHashCode();
+        int hashcode2 = HashCode.Combine(300, 15, 15);
+
+        Assert.IsTrue(hashcode1 != hashcode2);
```
Nit: casing
```suggestion
        int hashCode1 = angle1.GetHashCode();
        int hashCode2 = HashCode.Combine(300, 15, 15);

        Assert.IsTrue(hashCode1 != hashCode2);
```

Keboo commented on src/Chapter09.Tests/Listing09.23.OverridingGetHashCodeUsingValueTuples.Tests.cs:
```diff
@@ -0,0 +1,17 @@
+        Angle angle1 = new Angle(120, 20, 20, "name");
+        var testTuple = (120, 20, 20);
+
+        int hashcode1 = angle1.GetHashCode();
+        int hashcode2 = testTuple.GetHashCode();
+
+        Assert.AreEqual(hashcode1, hashcode2);
```
nit: Casing
```suggestion
        int hashCode1 = angle1.GetHashCode();
        int hashCode2 = testTuple.GetHashCode();

        Assert.AreEqual(hashCode1, hashCode2);
```

Keboo commented on src/Chapter09.Tests/Listing09.33.WritingEnumValueidentifierToTraceBuffers.Tests.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_33.Tests;
+
+enum ConnectionState
+{
+    Disconnected,
+    Connecting,
+    Connected,
+    Disconnecting
+}
```
We should not need to recreate this enum for the tests
```suggestion
```

Keboo commented on src/Chapter09.Tests/Listing09.33.WritingEnumValueidentifierToTraceBuffers.Tests.cs:
```diff
@@ -0,0 +1,34 @@
+
+        Trace.Listeners.Add(traceListener);
+
+        System.Diagnostics.Trace.WriteLine(
+            $"The connection is currently {ConnectionState.Disconnecting}");
```
This line should be invoking the Main method in the actual listing. As it stands now, this test is not actually calling the listing code.

Keboo commented on src/Chapter09/Chapter09.csproj:
```diff
@@ -1,11 +1,26 @@
+    <None Include="Listing09.13.RecordStructInitialization.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="NPOI" Version="2.6.1" />
```
Are we using something from the NPOI library? The book contents aim to avoid taking dependencies on outside libraries (with the exception of the tests)

danOIntellitect replied:
Has this comment been addressed as well as the warning excludes?

Keboo commented on src/Chapter09/Chapter09.csproj:
```diff
@@ -1,11 +1,26 @@
   <PropertyGroup>
     <ProductName>Chapter09</ProductName>
+    <ExpectedWarnings>$(ExpectedWarnings)</ExpectedWarnings>
+    <!-- CS0178: Invalid rank specifier: expected ',' or ']' -->
+    <!-- CS1513: } expected -->
+    <!-- CS1514: { expected -->
+    <!-- CS0826: No best type found for implicitly-typed array -->
+    <IgnoredWarnings>$(IgnoredWarnings),CS0178,CS1513,CS1514,CS0826</IgnoredWarnings>
```
Do we need to exclude these warnings? I would expect that we probably want to suppress them on a file by file basis rather than at the project level.

danOIntellitect replied:
Just bumping this comment

Keboo commented on src/Chapter09/Listing09.12.RecordConstructors.cs:
```diff
@@ -0,0 +1,14 @@
+    public Angle(
+        string degrees, string minutes, string seconds)
+        : this(int.Parse(degrees),
+              int.Parse(minutes), int.Parse(seconds))
```
nit: I realize this is how it was in the book, but I think stacking these calls makes the example cleaner. 
```suggestion
               int.Parse(minutes), 
               int.Parse(seconds))
```

Keboo commented on src/Chapter09/Listing09.18.OverridingEqualsAndNotEquals.cs:
```diff
@@ -0,0 +1,117 @@
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_18;
+
+[CompilerGenerated]
```
Except for the examples where we are attempting to show what the compiler has generated, we should drop this attribute.
```suggestion
```

Keboo commented on src/Chapter09/Listing09.19.NonMatchingSignatures.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_19;
+
+[CompilerGenerated]
```
```suggestion
```

Keboo commented on src/Chapter09/Listing09.20.ImplementingValueEqualityOnCustomTypes.cs:
```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections.Generic;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_20
```
I would make this a file scoped namespace.

Keboo commented on src/Chapter09/Listing09.21.ImplementStronglyTypedEquals.cs:
```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections.Generic;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_21
```
I would make this a file scoped namespace.

Keboo commented on src/Chapter09/Listing09.20.ImplementingValueEqualityOnCustomTypes.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_20
+{
+    public class Foo : IEquality<Foo>
```
Do the examples use `IEquality<T>`? I would have expected them to use the built in `IEquatable<T>`

elizabeth-pauley replied:
Yes, the book says to implement the `IEquality<T>` interface. There's a footnote attached to it that says it is a generic interface discussed in Chapter 12. I looked through Chapter 12 and couldn't find any mention of `IEquality<T>`, so I just made something of my own. Thoughts?

Keboo commented on src/Chapter09/Listing09.21.ImplementStronglyTypedEquals.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter09.Listing09_21
+{
+    public class Foo : IEquality<Foo>
```
Do the examples use `IEquality<T>`? I would have expected them to use the built in `IEquatable<T>`

elizabeth-pauley replied:
Same as above!

Keboo commented on src/Chapter09/Listing09.22.OverridingGetHashCodeUsingCombine.cs:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    #region INCLUDE
+    public override int GetHashCode() =>
+    HashCode.Combine(Degrees,
+        Minutes, Seconds);
```
nit: Spacing. I realize this is how it is in the book, but I think a little space makes it a bit easier to read.
```suggestion
    public override int GetHashCode() =>
        HashCode.Combine(Degrees, Minutes, Seconds);
```

Keboo commented on src/Chapter09/Listing09.22.OverridingGetHashCodeUsingCombine.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        if (other == null) return false;
+        return
+        EqualityComparer<int>.Default.Equals(Degrees, other.Degrees)
+        && EqualityComparer<int>.Default.Equals(Minutes, other.Minutes)
+        && EqualityComparer<int>.Default.Equals(Seconds, other.Seconds)
+        && EqualityComparer<string>.Default.Equals(Name, other.Name);
```
```suggestion
            EqualityComparer<int>.Default.Equals(Degrees, other.Degrees)
            && EqualityComparer<int>.Default.Equals(Minutes, other.Minutes)
            && EqualityComparer<int>.Default.Equals(Seconds, other.Seconds)
            && EqualityComparer<string>.Default.Equals(Name, other.Name);
```

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.02.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.02.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.06.EquivalentRecordClassCode.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.14.RecordDeconstructors.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.15.RecordDeconstructorsAndPatternMatching.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09/Listing09.22.OverridingGetHashCodeUsingCombine.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09/Listing09.22.OverridingGetHashCodeUsingCombine.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Changed Listing 21 to be a file scoped namespace

Keboo commented on src/Chapter09/Chapter09.csproj:
```diff
@@ -1,11 +1,26 @@
-</Project>
+  <ItemGroup>
+    <None Include="Listing09.13.RecordStructInitialization.cs" />
+  </ItemGroup>
```
These two ItemGroups can be combined into one.
```xml
  <ItemGroup>
    <Compile Remove="Listing09.13.RecordStructInitialization.cs" />
    <None Include="Listing09.13.RecordStructInitialization.cs" />
  </ItemGroup>
```

elizabeth-pauley replied:
Makes sense, changing that now!

elizabeth-pauley committed:
Chapter 9 updates

Changed a few pieces of code to more accurately match what the book originally had

elizabeth-pauley committed:
Edits to Chapter 9 per pull request conversation

Keboo commented on src/Chapter09.Tests/Listing09.14.RecordDeconstructors.Tests.cs:
```diff
@@ -0,0 +1,21 @@
+
+        angle.Deconstruct(out int degrees, out int minutes, out int seconds);
+
+        Assert.AreEqual(degrees, degrees);
+        Assert.AreEqual(minutes, minutes);
+        Assert.AreEqual(seconds, seconds);
```
Something is still off here. You are comparing values against themselves. Typically you will want to have the "expectec" (first parameter) be the same value from the "arrange" portion of the test

Keboo replied:
This one still doesn't look to be resolved. To be clear, the issue is `Assert.AreEqual(degree, degrees);` is just comparing `degrees` with itself (where it will always be true). We should be comparting it against the values that were passed above. Something more like: `Assert.AreEqual(90, degrees)`

elizabeth-pauley committed:
Modified test for listing 9.15

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.22.OverridingGetHashCodeUsingCombine.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09/Listing09.19.NonMatchingSignatures.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09/Listing09.18.OverridingEqualsAndNotEquals.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09/Listing09.12.RecordConstructors.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.33.WritingEnumValueidentifierToTraceBuffers.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
More updates to Chapter 9 tests

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.22.OverridingGetHashCodeUsingCombine.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Update src/Chapter09.Tests/Listing09.23.OverridingGetHashCodeUsingValueTuples.Tests.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

elizabeth-pauley committed:
Updated listing 9.14

danOIntellitect reviewed (approved):
LGTM as long as those comments that Kevin left were addressed. Let me know if the comments from him don't make sense

danOIntellitect committed:
fix: Addressed excluded warning and unsed reference
